### PR TITLE
Terra localnet

### PR DESCRIPTION
**Pull request by expertdicer**
## Summary of changes

Localrelayer is a local testing environment composed of three LocalTerra instances connected by a relayer. Localrelayer comes preconfigured with opinionated, sensible defaults for a standard testing environment.

### Presequite

Ensure you have docker and docker-compose installed:

```sh
# Docker
sudo apt-get remove docker docker-engine docker.io
sudo apt-get update
sudo apt install docker.io -y

# Docker compose
sudo apt install docker-compose -y
```

### Deploy

Build a local docker image with current changes

```bash
make build
```

Start the testing environment:

```bash
make start
```

> If you don't want the logs, you can start in detached mode with the following command:
> 
> `make startd`

Check that everything is running:

```bash
docker ps
```

Expected output:

```bash
❯ docker ps
CONTAINER ID   IMAGE                          COMMAND                  CREATED              STATUS         PORTS                                                                                   NAMES
318c89d3015f   informalsystems/hermes:1.1.0   "/home/hermes/setup.…"   About a minute ago   Up 2 seconds   0.0.0.0:3000->3000/tcp                                                                  localrelayer-hermes-1
d90ec29c7a6f   local:terra                  "/terra/setup.sh"      About a minute ago   Up 3 seconds   26656/tcp, 0.0.0.0:41317->1317/tcp, 0.0.0.0:49090->9090/tcp, 0.0.0.0:46657->26657/tcp   localrelayer-localterra-c-1
d90ec29c7a6f   local:terra                  "/terra/setup.sh"      About a minute ago   Up 3 seconds   26656/tcp, 0.0.0.0:31317->1317/tcp, 0.0.0.0:39090->9090/tcp, 0.0.0.0:36657->26657/tcp   localrelayer-localterra-b-1
e36cead49a07   local:terra                  "/terra/setup.sh"      About a minute ago   Up 3 seconds   0.0.0.0:1317->1317/tcp, 0.0.0.0:9090->9090/tcp, 0.0.0.0:26657->26657/tcp, 26656/tcp     localrelayer-localterra-a-1
```

### Testing

Run `IBC-Hooks` test:
```bash
sh ./tests/ibc-hooks/test_hooks.sh
```

Run `Packet Forward Middleware` test:
```bash
sh ./tests/ibc-hooks/test_pfm.sh
```

<Describe summary of major changes here>

## Report of required housekeeping

- [ ] Github issue OR spec proposal link
- [ ] Wrote tests
- [ ] Updated API documentation (client/lcd/swagger-ui/swagger.yaml)
- [ ] Added a relevant changelog entry: clog add [section] [stanza] [message]

----

## (FOR ADMIN) Before merging

- [ ] Added appropriate labels to PR
- [ ] Squashed all commits, uses message "Merge pull request #XYZ: [title]" (coding standards)
- [ ] Confirm added tests are consistent with the intended behavior of changes
- [ ] Ensure all tests pass

### CODE REVIEW

**Comment by nghuyenthevinh2000** [comment]
can we break down this pull request into multiple smaller items, reviewing 4000+ lines of code is not manageable

**Comment by expertdicer** [comment]
> can we break down this pull request into multiple smaller items, reviewing 4000+ lines of code is not manageable

I have reduced to 2000+

**Comment by expertdicer** [comment]
1000+ now

File: app/keepers/keepers.go [hunk 1]
```diff
@@ -375,14 +377,16 @@ func NewAppKeepers(
 		appCodec,
 		appKeepers.keys[ibctransfertypes.StoreKey],
 		appKeepers.GetSubspace(ibctransfertypes.ModuleName),
-		appKeepers.IBCHooksWrapper,
```
[the remark this answers is not in the document]
Comment (expertdicer) [reply app/keepers/keepers.go:378]: yes. The correct ics4 wrapper should be `&appKeepers.ForwardKeeper`

**Comment by nghuyenthevinh2000** [comment]
Thank you for your pull request. Given the current ambivalence within the community regarding PPJ, we will temporarily suspend the reviewing process until a well-defined plan for PPJ, which includes the reviewing process and delineates the shares of responsibility, is approved through governance.

**Pull request closed without merging**